lib/cluster/subscriber.js

```diff
@@ -0,0 +1,699 @@
+    this.socket.subscribe();
+    this.heartbeatTimer = setInterval(
+      () => this.checkHeartbeat(),
+      this.heartbeatDelay);
+
+    this.listen();
+  }
+
+  /**
+   * Starts subscribing to other nodes
+   * Do NOT wait this method: it's an infinite loop, it's not meant to ever
+   * return (unless the remote node has been evicted)
+   */
+  async listen () {
+    while (this.state !== stateEnum.EVICTED) {
+      let topic;
+      let data;
+
+      try {
+        [topic, data] = await this.socket.receive();
+      }
+      catch (e) {
+        if (this.state !== stateEnum.EVICTED) {
+          await this.evictNode({
+            broadcast: true,
+            reason: e.message,
+          });
+        }
+
+        return;
+      }
+
+      // do nothing is the node was evicted even if we already check
+      // for this in processData()
+      if (this.state === stateEnum.EVICTED) {
+        return;
+      }
+
+      if (this.state !== stateEnum.BUFFERING) {
+        await this.processData(topic.toString(), data);
+      }
+      else {
+        this.buffer.push([topic.toString(), data]);
+      }
+    }
+  }
+
+  /**
+   * Plays all buffered sync messages, and switches this subscriber state from
+   * BUFFERING to SANE.
+   *
+   * @param  {Long} lastMessageId
+   * @return {void}
+   */
+  async sync (lastMessageId) {
+    this.lastMessageId = lastMessageId;
+
+    // copy the buffer for processing: new sync messages might be buffered
+    // while we apply older messages with async functions
+    while (this.buffer.length > 0) {
+      const _buffer = this.buffer;
+      this.buffer = [];
+
+      for (let i = 0; i < _buffer.length; i++) {
+        await this.processData(_buffer[i][0], _buffer[i][1]);
+      }
+    }
+
+    this.state = stateEnum.SANE;
+  }
+
+  /**
+   * Decodes an incoming message, and dispatches it.
+   * The topic name must match a protobuf message.
+   *
+   * /!\ This method MUST NEVER THROW
+   * It's awaited by this.listen(), which cannot be awaited (infinite loop),
+   * and it also cannot afford to attach a rejection handler everytime a message
+   * is received to prevent clogging the event loop with unnecessary promises
+   *
+   * @param  {string} topic
+   * @param  {Buffer} data
+   * @return {void}
+   */
+  async processData (topic, data) {
+    if (this.state === stateEnum.EVICTED) {
+      return;
+    }
+
+    const decoder = this.protoroot.lookup(topic);
+
+    if (decoder === null) {
+      await this.evictNode({
+        broadcast: true,
+        reason: `received an invalid message from ${this.remoteNodeId} (unknown topic "${topic}")`,
+      });
+      return;
+    }
+
+    const message = decoder.toObject(decoder.decode(data));
+
+    if (!await this.validateMessage(message)) {
+      return;
+    }
+
+    try {
+      await this.handlers[topic].call(this, message);
+    }
+    catch (e) {
+      // @todo can't the node evict itself to save a round of heartbeat?
```

Comment:
Can't we broadcast a node eviction message to save a round of heartbeat to other nodes?

Reply:
yes you're right for both comments

Reply:
this will also provide insight as to why a node shuts down

Reply:
Other nodes won't have to wait for a round of heartbeat: there is a shutdown handler broadcasting a `NodeShutdown` message to all other nodes.

Reply:
@Aschen > fixed. Nodes that shut down themselves because of a malfunction now first send an eviction message, with the shutdown reason in it, for activity tracking purposes.

I didn't bother doing the same when an eviction occurs due to a network split, since by definition, 1/ the cluster cannot receive the eviction message, and 2/ every other nodes will conclude on the same eviction and will log it in their activity tracking themselves.

I also adapted the unit tests accordingly, removed the custom kuzzle mock you wrote, and replaced it with our standard one. Which shed light on a bug: when handling a dump request, I incorrectly called `kuzzle.janitor.dump` instead of `kuzzle.dump`: since the standard mock is always up-to-date, this made the bug apparent :wink: